public/js/all.js

```diff
@@ -114,9 +114,33 @@ function replaceAll(find, replace, str) {
   return str.replace(new RegExp(find, 'g'), replace);
 }
 
+var cardsCache = {};
```

Comment:
I presume it wouldn't be difficult to populate this from the body of a page? For example, there's no reason not to preload the achievement tooltips on a game page as all the necessary data is already shown on the page itself.

Reply:
> all the necessary data

True, yet sounds like premature optimization to me (ironically). Not all of said data will be available at all times as we might want to move around sections and context of pages.
Making it dependent on the context of the currently viewed page introduces complexity i'd rather get rid of/not introduce. Rewriting it all to have it prepared on page load seems overly complicated to me. I'm open for suggestions however, i cannot claim to have it all thought through.

To me, loading tooltips cards should be a separate aspect of the system at all times; able to be cached and rendered separately from the rest of the site.

I'm aware that there most of the data is already fetched at the time the tooltips could be rendered. Still, not assuming that all data is available, but fetching it when needed seems wiser to me. We should look for loading _less_ data for rendering the initial data instead. This PR opens some opportunities for that.

Reply:
> Not all of said data will be available at all times

Obviously, if the information isn't available, avoiding the query to fetch it would be desirable. But in cases where it is available, wouldn't it be desirable to avoid the extra query and the extra round trip to the server?

For the game page, the data is available - including the unlock timestamps:
![image](https://user-images.githubusercontent.com/32680403/192181409-1dc3eac1-06b7-4246-8322-b4c5e7e3e08b.png)

I agree that it's unlikely that a user would need all of the tooltips, but if the data is available, why not send it on the first response and avoid the extra round trips.

For the game page, the tooltip don't even seem necessary. They're just redundant, which supports my assertation that all the necessary data is already available. In this case, it might be better to remove the tooltip entirely.

For something like the achievement grids on the profile page, the tooltip becomes much more important as there is no text displayed. The query to build the recent games grid already fetches all the information necessary to generate the tooltip (achievement title, description, points, game title, unlock date):
https://github.com/RetroAchievements/RAWeb/blob/62b1e3216fa076b5cde344be14f02e516db78f34/lib/database/player-achievement.php#L232

So why not use it?

Reply:
> why not send it on the first response and avoid the extra round trips
> already fetches all the information necessary
> So why not use it?

Normally i'd absolutely agree and we can surely revisit that option at another time. It's an odd situation where the database is not the only bottleneck for response performance.

I think it introduces code for optimizing something that is not that much of an issue. Maybe I'm thinking of the wrong thing there and we should give it a try (in a separate step)-

The round trips currently aren't as expensive as rendering time. Even the suggested combination of icon and text queries aren't as much of a problem in comparison to the excessive pre-rendering of multiple hundreds tooltips on some pages.
We have forum topics that contain an almost uncapped amount of references to games/achievements/users which not only resulted in a lot of (sometimes expensive) queries but also pushed the rendering time to an excessive amount.

The server's cpu load is fairly evenly distributed between PHP and mysql atm - PHP should not take that much time rendering but leave as much resources to the database as possible (whilst we also reduce its own usage).

Reply:
> already fetches all the information necessary

I think that we actually could reduce the amount of data fetched, there's certainly some overfetching happening in some areas. I'll try to identify those areas at a later time.

Reply:
> We have forum topics that contain an almost uncapped amount of references to games/achievements/users which not only resulted in a lot of (sometimes expensive) queries but also pushed the rendering time to an excessive amount.

I'm not suggesting we do it everywhere. Forums wouldn't have that data available, and I agree that pre-fetching it there is overhead that should be avoided. 

I'm just suggesting that in places where the data _is_ available, we could use it and avoid the extra server hit(s)/queries.

Reply:
I understand :) Personally, I'd rather take the occasional extra hits (which are only done once for each individual tooltip and most of the time not if no interaction happens) than introducing complexity by conditionally rendering them in some places and not in others. It makes it more rigid, too, as we'd have to _assure_ that all data will be kept available in those places where we have the data available.

I'll just give it a try to see what it could look like.

Reply:
Working on a separate PR for the optional tooltip data.